sdk/ml/azure-ai-ml/azure/ai/ml/_azure_environments.py

```diff
@@ -189,3 +192,130 @@ def _resource_to_scopes(resource):
+
+def _get_clouds_by_metadata_url(metadata_url, timeout=DEFAULT_TIMEOUT):
+    """Get all the clouds by the specified metadata url
+
+        :return: list of the clouds
+    """
+    try:
+        import requests
+        module_logger.debug('Start : Loading cloud metatdata from the url specified by {0}'.format(metadata_url))
+        with requests.get(metadata_url, timeout=timeout) as meta_response:
+            arm_cloud_dict = meta_response.json()
+            cli_cloud_dict = _convert_arm_to_cli(arm_cloud_dict)
+            module_logger.debug('Finish : Loading cloud metatdata from the url specified by {0}'.format(metadata_url))
+            return cli_cloud_dict
+    except Exception as ex:  # pylint: disable=broad-except
+        module_logger.warning("Error: Azure ML was unable to load cloud metadata from the url specified by {0}. {1}. "
+                        "This may be due to a misconfiguration of networking controls. Azure Machine Learning Python SDK "
+                        "requires outbound access to Azure Resource Manager. Please contact your networking team to configure "
+                        "outbound access to Azure Resource Manager on both Network Security Group and Firewall. "
+                        "For more details on required configurations, see "
+                        "https://docs.microsoft.com/azure/machine-learning/how-to-access-azureml-behind-firewall.".format(
+            metadata_url, ex))
+
+def _convert_arm_to_cli(arm_cloud_metadata_dict):
+    cli_cloud_metadata_dict = {}
+    for cloud in arm_cloud_metadata_dict:
+        cli_cloud_metadata_dict[cloud['name']] = {
+            EndpointURLS.AZURE_PORTAL_ENDPOINT: cloud["portal"],
+            EndpointURLS.RESOURCE_MANAGER_ENDPOINT: cloud["resourceManager"],
+            EndpointURLS.ACTIVE_DIRECTORY_ENDPOINT: cloud["authentication"]["loginEndpoint"],
+            EndpointURLS.AML_RESOURCE_ID: cloud["resourceManager"],
+            EndpointURLS.STORAGE_ENDPOINT: cloud["suffixes"]["storage"]
+        }
+    return cli_cloud_metadata_dict
+
+def _get_cloud(cloud_name):
+    return next((data for name,data in _get_all_clouds().items() if name == cloud_name), None)
+
+
+def _get_all_clouds():
+    # Start with the hard coded list of clouds in this file
+    all_clouds = {}
+    all_clouds.update(_environments)
+    # Get configs from the config file
+    config = configparser.ConfigParser()
+    try:
+        config.read(CLOUD_CONFIG_FILE)
+    except configparser.MissingSectionHeaderError:
+        os.remove(CLOUD_CONFIG_FILE)
+        module_logger.warning("'%s' is in bad format and has been removed.", CLOUD_CONFIG_FILE)
+    for section in config.sections():
+        all_clouds[section] = dict(config.items(section))
+    # Now do the metadata URL
+    arm_url = os.environ[_ARM_METADATA_URL_ENV_NAME] if _ARM_METADATA_URL_ENV_NAME in os.environ else _DEFAULT_ARM_URL
+    all_clouds.update(_get_clouds_by_metadata_url(arm_url))
```

Comment:
Curious is this will throw an error in case of the metadata cannot be retrieved. for any reason?